cmd/zoekt-dynamic-indexserver/main.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2016 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This program manages a zoekt dynamic indexing deployment:
+// * listens to indexing commands
+// * reindexes specified repositories
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"time"
+)
+
+func loggedRun(cmd *exec.Cmd) (err error) {
+	outBuf := &bytes.Buffer{}
+	errBuf := &bytes.Buffer{}
+	cmd.Stdout = outBuf
+	cmd.Stderr = errBuf
+
+	log.Printf("run %v", cmd.Args)
+	if err = cmd.Run(); err != nil {
+		log.Printf("command %s failed: %v\nOUT: %s\nERR: %s",
+			cmd.Args, err, outBuf.String(), errBuf.String())
+		err = fmt.Errorf("command %s failed: %v", cmd.Args, err)
+	}
+
+	return
+}
+
+type Options struct {
+	indexTimeout time.Duration
+	dataDir      string
+	indexDir     string
+	repoDir      string
+	listen       string
+}
+
+func (o *Options) createMissingDirectories() {
+	for _, s := range []string{o.dataDir, o.indexDir, o.repoDir} {
+		if _, err := os.Stat(s); err == nil {
+			continue
+		}
+
+		if err := os.MkdirAll(s, 0o755); err != nil {
+			log.Fatalf("MkdirAll %s: %v", s, err)
+		}
+	}
+}
+
+type indexRequest struct {
+	CloneURL string // TODO: Decide if tokens can be in the URL or if we should pass separately
+	RepoID   uint32
+}
+
+// This function is declared as var so that we can stub it in test
+var executeCmd = func(ctx context.Context, name string, arg ...string) (err error) {
+	cmd := exec.CommandContext(ctx, name, arg...)
+	cmd.Stdin = &bytes.Buffer{}
+	err = loggedRun(cmd)
+
+	return
+}
+
+func indexRepository(opts Options, req indexRequest) (response map[string]any, err error) {
+	ctx, cancel := context.WithTimeout(context.Background(), opts.indexTimeout)
```

Comment:
to clarify how you would use the request context, even though it is likely better to not cancel slow indexes (unless internally you can have long lasting requests) is something like
```suggestion
func indexRepository(ctx context.Context, opts Options, req indexRequest) (response map[string]any, err error) {
	ctx, cancel := context.WithTimeout(ctx, opts.indexTimeout)
```

then at the call site of indexRepository you can use the http.Request.Context